### PR TITLE
History tracking

### PR DESCRIPTION
**Pull request by gregorjerse**

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -96,6 +96,7 @@
             "isort>=5.12.0",
             "daphne>=4.0.0",
             "django-stubs>=4.2.4",
+            "djangorestframework-stubs[compatible-mypy]>=3.14.2",
```
Comment (dblenkus) [inline setup.py:99]: This version should probably be the same as `djangorestframework`, so `~=3.14.0`.

File: resolwe/flow/models/base.py [hunk 2]
```diff
@@ -12,6 +12,8 @@
 
 # Maximum number of slug-generation retries.
 MAX_SLUG_RETRIES = 10
+SLUG_LENGTH = 100
```
Comment (dblenkus) [inline resolwe/flow/models/base.py:15]: Prepend these constants with `MAX_`.

File: resolwe/flow/models/history.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+from django.db import models
+from django.db.backends.postgresql.psycopg_any import DateTimeTZRange
+
+from .base import NAME_LENGTH, SLUG_LENGTH
+from .collection import Collection
+from .data import Data
+from .history_manager import HistoryMixin, history_manager
+
+
+class BaseHistory(models.Model):
```
Comment (dblenkus) [inline resolwe/flow/models/history.py:18]: Mark this class as abstract: https://docs.djangoproject.com/en/4.2/topics/db/models/#abstract-base-classes

File: resolwe/flow/models/history_manager.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+                    if current_value != new_value:
+                        tracker.process_change(instance, field_name)
+            except RuntimeError:
+                tracker.process_change(instance, field_name)
+
+
+history_manager = HistoryManager()
+
+
+def post_delete_signal(sender, instance, **kwargs):
```
Comment (dblenkus) [inline resolwe/flow/models/history_manager.py:106]: Move this to the `delete` method.

File: setup.py (hunk 1 above)
Comment (gregorjerse) [reply setup.py:99]: Not sure: `djangorestframework` has no `3.14.2` release, but unifying them will not hurt.

File: resolwe/flow/models/history_manager.py (hunk 4 above)
Comment (gregorjerse) [reply resolwe/flow/models/history_manager.py:106]: I was just about to do that when a paragraph

```The delete() method does a bulk delete and does not call any delete() methods on your models. It does, however, emit the [pre_delete](https://docs.djangoproject.com/en/4.2/ref/signals/#django.db.models.signals.pre_delete) and [post_delete](https://docs.djangoproject.com/en/4.2/ref/signals/#django.db.models.signals.post_delete) signals for all deleted objects (including cascaded deletions).```

from Django documentation changed my mind. However, I do not like to depend on signals to much so I will move this to a `delete` method. And keep in mind that bulk delete is not a good idea.

File: resolwe/flow/models/base.py (hunk 2 above)
Comment (gregorjerse) [reply resolwe/flow/models/base.py:15]: Done.

File: resolwe/flow/models/history.py (hunk 3 above)
Comment (gregorjerse) [reply resolwe/flow/models/history.py:18]: This is not possible: I have a reference to this class from `SlugHistory` for example. This enables me to use the same class to track history of multiple objects.

**Review by dblenkus — APPROVED** [review]

**Pull request merged**